pkg/server/endpoints/bundle/acme_auth.go

```diff
@@ -0,0 +1,192 @@
+// simplifying assumptions based on our usage for the bundle endpoint. Namely,
+// it assumes there is going to be a single cache entry, since we only support
+// a single domain. It assumes PEM encoded blocks of data and strips out the
+// private key to be stored in the key manager instead of on disk with the rest
+// of the data.
+type ACMEConfig struct {
+	// DirectoryURL is the ACME directory URL
+	DirectoryURL string
+
+	// DomainName is the domain name of the certificate to obtain.
+	DomainName string
+
+	// CacheDir is the directory on disk where we cache certificates.
+	CacheDir string
+
+	// Email is the email address of the account to register with ACME
+	Email string
+}
+
+func ACMEAuth(log logrus.FieldLogger, km keymanager.KeyManager, config ACMEConfig) ServerAuth {
+	// The acme client already defaulting to Let's Encrypt if the URL is unset
+	// but we want it populated for logging purposes.
+	if config.DirectoryURL == "" {
+		config.DirectoryURL = acme.LetsEncryptURL
+	}
+
+	return &acmeAuth{
+		m: &autocert.Manager{
+			Prompt: func(tosURL string) bool {
+				log.WithFields(logrus.Fields{
+					"directory_url": config.DirectoryURL,
+					"tos_url":       tosURL,
+					"email":         config.Email,
+				}).Info("ACME Terms of Service accepted")
+				return true
+			},
+			Email:      config.Email,
+			Cache:      autocert.DirCache(config.CacheDir),
+			HostPolicy: autocert.HostWhitelist(config.DomainName),
+			Client: &acme.Client{
+				DirectoryURL: config.DirectoryURL,
+				UserAgent:    "SPIRE-" + version.Version(),
+			},
+			KeyStore: &acmeKeyStore{
+				log: log,
+				km:  km,
+			},
+		},
+	}
+}
+
+type acmeAuth struct {
+	m *autocert.Manager
+}
+
+func (a *acmeAuth) GetTLSConfig() *tls.Config {
+	return a.m.TLSConfig()
+}
+
+type acmeKeyStore struct {
+	log logrus.FieldLogger
+	km  keymanager.KeyManager
+}
+
+func (ks *acmeKeyStore) GetPrivateKey(ctx context.Context, id string) (crypto.Signer, error) {
+	keyID := acmeKeyPrefix + id
+
+	resp, err := ks.km.GetPublicKey(ctx, &keymanager.GetPublicKeyRequest{
+		KeyId: keyID,
+	})
+	if err != nil {
+		return nil, errs.Wrap(err)
+	}
+	if resp.PublicKey == nil {
+		return nil, nil
+	}
+
+	signer, err := ks.signer(keyID, resp.PublicKey)
+	if err != nil {
+		return nil, err
+	}
+
+	if isACMEAccountKey(id) {
+		// The account key is used to sign JWT tokens and needs to sign things
+		// differently. Unfortunately, github.com/x/crypto/acme does not
+		// properly handle crypto.Signers that aren't of type *ecdsa.PrivateKey
+		// so we have to wrap it here.
+		signer = jwtSigner{Signer: signer}
+	}
+	return signer, nil
+}
+
+func (ks *acmeKeyStore) NewPrivateKey(ctx context.Context, id string, keyType autocert.KeyType) (crypto.Signer, error) {
+	keyID := acmeKeyPrefix + id
+
+	var kmKeyType keymanager.KeyType
+	switch keyType {
+	case autocert.RSA2048:
+		kmKeyType = keymanager.KeyType_RSA_2048
+	case autocert.EC256:
+		kmKeyType = keymanager.KeyType_EC_P256
```

Comment:
The autocert package is intended to light up TLS clients who don't support on key type or the other. If a client that does not support EC connects, for example, it will request an RSA certificate for the domain. If a client that does support EC connects, it will request an EC certificate. It maintains both so it can serve both kinds of clients.